File: tasks/implementations/python/tasks/service_support.py
import json
import socket
import subprocess
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

DB_PATH = "/tmp/bench.sqlite"


def _sqlite(query: str) -> str:
    proc = subprocess.run(["sqlite3", DB_PATH, query], check=True, capture_output=True, text=True)
    return proc.stdout.strip()


def start_http_service(task: str, port: int) -> None:
    if task == "sqlite_crud":
        _sqlite("CREATE TABLE IF NOT EXISTS items (id INTEGER PRIMARY KEY, value INTEGER);")

    class Handler(BaseHTTPRequestHandler):
        def do_GET(self):
            parsed = urlparse(self.path)
            query = parse_qs(parsed.query)
            if parsed.path == "/health":
                self._respond(200, "text/plain", "ok")
                return
            if task == "simple_web_server" and parsed.path == "/":
                self._respond(200, "text/plain", "hello-benchmark")
                return
            if task in {"gui_calculator", "data_visualization", "basic_web_application"} and parsed.path == "/":
                body = (Path("/fixtures") / "ui" / f"{task}.html").read_text(encoding="utf-8")
                self._respond(200, "text/html; charset=utf-8", body)
                return
            if task == "rest_api" and parsed.path == "/item":
                item_id = int(query.get("id", ["0"])[0])
                self._respond_json({"id": item_id, "value": item_id * 2})
                return
            if task == "sqlite_crud":
                item_id = int(query.get("id", ["0"])[0])
                if parsed.path == "/create":
                    value = int(query.get("value", ["0"])[0])
                    _sqlite(f"INSERT OR REPLACE INTO items(id, value) VALUES({item_id}, {value});")
                    self._respond_json({"status": "created", "id": item_id})
                    return
                if parsed.path == "/read":
                    value = _sqlite(f"SELECT value FROM items WHERE id={item_id} LIMIT 1;")
                    self._respond_json({"id": item_id, "value": int(value) if value else None})
                    return
                if parsed.path == "/update":
                    value = int(query.get("value", ["0"])[0])
                    _sqlite(f"UPDATE items SET value={value} WHERE id={item_id};")
                    self._respond_json({"status": "updated", "id": item_id})
                    return
                if parsed.path == "/delete":
                    _sqlite(f"DELETE FROM items WHERE id={item_id};")
                    self._respond_json({"status": "deleted", "id": item_id})
                    return
            self._respond(404, "text/plain", "not-found")

        def log_message(self, *_args):
            return

        def _respond(self, status: int, content_type: str, body: str):
            payload = body.encode("utf-8")
            self.send_response(status)
            self.send_header("Content-Type", content_type)
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)

        def _respond_json(self, obj):
            self._respond(200, "application/json", json.dumps(obj))

    HTTPServer(("0.0.0.0", port), Handler).serve_forever()


def start_tcp_service(task: str, port: int) -> None:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server:
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server.bind(("0.0.0.0", port))
        server.listen()
        while True:
            conn, _ = server.accept()
            with conn:
                data = conn.recv(4096)
                if not data:
                    continue
                payload = data.decode("utf-8")
                response = ("chat:" + payload) if task == "chat_application" else payload
                conn.sendall(response.encode("utf-8"))

File: tasks/implementations/go/tasks/service_support.go
package main

import (
    "encoding/json"
    "fmt"
    "net"
    "net/http"
    "os"
    "os/exec"
    "strconv"
)

const dbPath = "/tmp/bench.sqlite"

func sqlite(query string) string {
    out, err := exec.Command("sqlite3", dbPath, query).CombinedOutput()
    if err != nil { panic(string(out)) }
    return string(bytesTrim(out))
}

func bytesTrim(value []byte) []byte {
    for len(value) > 0 && (value[len(value)-1] == '\n' || value[len(value)-1] == '\r' || value[len(value)-1] == ' ') {
        value = value[:len(value)-1]
    }
    return value
}

func startHTTPService(task string, port int) {
    if task == "sqlite_crud" {
        sqlite("CREATE TABLE IF NOT EXISTS items (id INTEGER PRIMARY KEY, value INTEGER);")
    }
    mux := http.NewServeMux()
    mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) { fmt.Fprint(w, "ok") })
    mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
        switch task {
        case "simple_web_server":
            fmt.Fprint(w, "hello-benchmark")
        case "rest_api":
            if r.URL.Path != "/item" { http.NotFound(w, r); return }
            id, _ := strconv.Atoi(r.URL.Query().Get("id"))
            json.NewEncoder(w).Encode(map[string]int{"id": id, "value": id * 2})
        case "gui_calculator", "data_visualization", "basic_web_application":
            if r.URL.Path != "/" { http.NotFound(w, r); return }
            body, _ := os.ReadFile("/fixtures/ui/" + task + ".html")
            w.Header().Set("Content-Type", "text/html; charset=utf-8")
            fmt.Fprint(w, string(body))
        case "sqlite_crud":
            id, _ := strconv.Atoi(r.URL.Query().Get("id"))
            switch r.URL.Path {
            case "/create":
                value, _ := strconv.Atoi(r.URL.Query().Get("value"))
                sqlite(fmt.Sprintf("INSERT OR REPLACE INTO items(id, value) VALUES(%d, %d);", id, value))
                json.NewEncoder(w).Encode(map[string]any{"status": "created", "id": id})
            case "/read":
                value := sqlite(fmt.Sprintf("SELECT value FROM items WHERE id=%d LIMIT 1;", id))
                if value == "" { json.NewEncoder(w).Encode(map[string]any{"id": id, "value": nil}) } else { parsed, _ := strconv.Atoi(value); json.NewEncoder(w).Encode(map[string]any{"id": id, "value": parsed}) }
            case "/update":
                value, _ := strconv.Atoi(r.URL.Query().Get("value"))
                sqlite(fmt.Sprintf("UPDATE items SET value=%d WHERE id=%d;", value, id))
                json.NewEncoder(w).Encode(map[string]any{"status": "updated", "id": id})
            case "/delete":
                sqlite(fmt.Sprintf("DELETE FROM items WHERE id=%d;", id))
                json.NewEncoder(w).Encode(map[string]any{"status": "deleted", "id": id})
            default:
                http.NotFound(w, r)
            }
        default:
            http.NotFound(w, r)
        }
    })
    http.ListenAndServe(fmt.Sprintf(":%d", port), mux)
}

func startTCPService(task string, port int) {
    listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
    if err != nil { panic(err) }
    for {
        conn, err := listener.Accept()
        if err != nil { continue }
        go func(c net.Conn) {
            defer c.Close()
            buf := make([]byte, 4096)
            n, _ := c.Read(buf)
            if n <= 0 { return }
            payload := string(buf[:n])
            if task == "chat_application" { c.Write([]byte("chat:" + payload)) } else { c.Write([]byte(payload)) }
        }(conn)
    }
}
